apps/arweave/src/ar_data_sync.erl

```diff
@@ -1078,6 +1079,7 @@ handle_cast({store_fetched_chunk, Peer, Byte, Proof} = Cast, State) ->
 	Offset = SeekByte - BlockStartOffset,
 	ValidateDataPathRuleset = ar_poa:get_data_path_validation_ruleset(BlockStartOffset,
 			get_merkle_rebase_threshold()),
+	IsLocalPeer = lists:member(Peer, State#sync_data_state.local_peers),
```

Comment:
Nice - elegant!

A couple points:
1. It's possible to query chunks from local_peers that *aren't* packed to our target packing format. In that case we don't need to validate, but we do still need to repack. Will that still happen in this flow?
2. When we are querying from a local_peer, ideally we'd skip the validate_proof call entirely.

Reply:
Basically, we need to call for `validate_proof` here exactly to check if the chunk is packed as we want, or we need to repack it. The function itself just takes the proof and does some calculations on it, so no long calls or other delays involved, should be fast.

Reply:
I don't think validate_proof checks if the chunk is packed correctly. I believe it just checks if it's packed or unpacked - if it's unpacked it can validate the chunk directly, if it's packed to any format it returns `need_unpacking`. It is `pack_and_store_chunk` that then checks if a valid chunk is packed correctly ad packs it as needed.

So maybe the only change needed is to swap `store_chunk` to `pack_and_store_chunk` - however in that case it might still be worth avoiding `validate_proof` mostly because I don't believe it's necessary. Even though performance may not be a concern, it adds a layer to the logic flow that needs to be  understood and maintained (e.g. the next person to see the code will wonder why validate_proof is called even though the result is ignored, and may spend time digging into the code, etc... alternatively it's possible someone adds a side effect to validate_proof or changes its return value which then impacts the "no validation needed" case, etc...)

Also, given the complexity of this logic, it would be good to also run some live tests (perhaps using debug output to confirm the correct behavior).

i.e.

1. sync from local peers where client requests chunks packed correctly. Chunk should be written without any repacking
2. sync from local peers where client requests `any` chunks, chunks should be repacked before being written.

Reply:
@JamesPiechota yes, I articulated it incorrectly, it does not check the packing, but it checks the proof, as the name implies :)
The thing is that to get all the proper arguments for storing the chunk I will need to duplicate half of the code that is doing the proof validation now. In ideal circumstances you're probably right, the proof validation is not needed, but in this case I think this is the best solution.
I will do all the tests, of course.